subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/OptionalDependenciesHandler.java

```diff
@@ -0,0 +1,113 @@
+
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class is responsible for maintaining the state of optional dependencies which are "pending".
+ * In other words, when an optional dependency is added to the graph, it is "pending" until a hard
+ * dependency for the same module is seen. As soon as a hard dependency is found, nodes that referred
+ * to the optional dependency are restarted.
+ *
+ * This class also makes a special case of the "optional" configuration, for backwards compatibility:
+ * the optional configuration used to store optional dependencies. But if we no longer resolve optional
+ * dependencies, then the optional configuration becomes effectively empty. To avoid this, we ignore the
+ * state of optional dependencies if they belong to the "optional" configuration.
+ */
+public class OptionalDependenciesHandler {
+    private final Map<ModuleIdentifier, PendingOptionalDependencies> optionalDependencies;
+    private final ImmutableModuleIdentifierFactory moduleIdentifierFactory;
+    private final DependencySubstitutionApplicator dependencySubstitutionApplicator;
+
+    private boolean started;
+    private boolean isOptionalConfiguration;
+    private List<PendingOptionalDependencies> noLongerOptional;
+
+    public OptionalDependenciesHandler(Map<ModuleIdentifier, PendingOptionalDependencies> optionalDependencies,
+                                       ImmutableModuleIdentifierFactory moduleIdentifierFactory,
+                                       DependencySubstitutionApplicator dependencySubstitutionApplicator) {
+
+        this.optionalDependencies = optionalDependencies;
+        this.moduleIdentifierFactory = moduleIdentifierFactory;
+        this.dependencySubstitutionApplicator = dependencySubstitutionApplicator;
+    }
+
+    private ModuleIdentifier toKey(DependencyState dependencyState) {
+        DependencySubstitutionApplicator.Application application = dependencySubstitutionApplicator.apply(dependencyState.getDependencyMetadata());
+        DependencySubstitutionInternal details = application.getResult();
+        if (details != null && details.isUpdated()) {
+            ComponentSelector target = details.getTarget();
+            if (target instanceof ModuleComponentSelector) {
+                ModuleComponentSelector selector = (ModuleComponentSelector) target;
+                return moduleIdentifierFactory.module(selector.getGroup(), selector.getModule());
+            }
+        }
+        return dependencyState.getModuleIdentifier();
+    }
+
+    boolean maybeAddAsOptionalDependency(NodeState node, DependencyState dependencyState) {
+        ModuleIdentifier key = toKey(dependencyState);
+        PendingOptionalDependencies pendingOptionalDependencies = optionalDependencies.get(key);
+
+        if (dependencyState.getDependencyMetadata().isOptional() && !isOptionalConfiguration) {
+            if (pendingOptionalDependencies == null || pendingOptionalDependencies.isPending()) {
```

Comment:
⭕️ It seems like replacing the `Map` with a more explicit class like `PendingOptionalModules` with methods like `isPending(id)` and `getPendingDependencies(id)` and `addDependency(id, dependency)` would make this code clearer.